Java 21: Unit test failed when running flow main branch

### Description of the bug

flow unit test failed with running Java 21 


```
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.035 s <<< FAILURE! -- in com.vaadin.flow.i18n.DefaultInstantiatorI18NTest
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10654&logView=flowAware&linesState=10654)      [ERROR] com.vaadin.flow.i18n.DefaultInstantiatorI18NTest.translationFileOnClasspath_instantiateDefaultI18N -- Time elapsed: 0.008 s <<< FAILURE!
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10655&logView=flowAware&linesState=10655)      java.lang.AssertionError
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10656&logView=flowAware&linesState=10656)        at org.junit.Assert.fail(Assert.java:87)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10657&logView=flowAware&linesState=10657)        at org.junit.Assert.assertTrue(Assert.java:42)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10658&logView=flowAware&linesState=10658)        at org.junit.Assert.assertTrue(Assert.java:53)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10659&logView=flowAware&linesState=10659)        at com.vaadin.flow.i18n.DefaultInstantiatorI18NTest.translationFileOnClasspath_instantiateDefaultI18N(DefaultInstantiatorI18NTest.java:95)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10660&logView=flowAware&linesState=10660)        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10661&logView=flowAware&linesState=10661)        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10662&logView=flowAware&linesState=10662)        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
[09:16:53 ](https://bender.vaadin.com/buildConfiguration/JDKs/453258?buildTab=log&focusLine=10663&logView=flowAware&linesState=10663)        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
```

### Expected behavior

build is working

### Minimal reproducible example

- download the currently java 21 version 
  - `https://download.java.net/java/GA/jdk21.0.1/415e3f918a1f4062a0074a2794853d0d/12/GPL/openjdk-21.0.1_linux-x64_bin.tar.gz`

- `mvn -e -B -V -T 6 -q -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn test -am -pl flow-html-components,flow-dnd,flow-plugins`

### Versions

- Vaadin / Flow version: 24.3-SNAPSHOT
- Java version: 21.0.1

--- a/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
@@ -28,6 +28,7 @@ import java.util.Locale;
 import java.util.ResourceBundle;
 import java.util.concurrent.atomic.AtomicReference;
 
+import net.jcip.annotations.NotThreadSafe;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -42,6 +43,7 @@ import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.VaadinService;
 
+@NotThreadSafe
 public class DefaultInstantiatorI18NTest {
 
     @Rule

--- a/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
@@ -18,7 +18,6 @@ package com.vaadin.flow.i18n;
 
 import java.io.File;
 import java.io.IOException;
-import java.lang.reflect.Field;
 import java.net.URL;
 import java.net.URLClassLoader;
 import java.nio.charset.StandardCharsets;
@@ -26,7 +25,6 @@ import java.nio.file.Files;
 import java.nio.file.StandardOpenOption;
 import java.util.Locale;
 import java.util.ResourceBundle;
-import java.util.concurrent.atomic.AtomicReference;
 
 import net.jcip.annotations.NotThreadSafe;
 import org.junit.After;
@@ -40,6 +38,7 @@ import org.mockito.Mockito;
 
 import com.vaadin.flow.di.DefaultInstantiator;
 import com.vaadin.flow.di.Lookup;
+import com.vaadin.flow.server.I18NProviderTest;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.VaadinService;
 
@@ -51,7 +50,6 @@ public class DefaultInstantiatorI18NTest {
 
     private File translations;
     private ClassLoader urlClassLoader;
-    private DefaultInstantiator instantiator;
 
     @Before
     public void init() throws IOException {
@@ -67,13 +65,7 @@ public class DefaultInstantiatorI18NTest {
     @After
     public void cleanup() throws NoSuchFieldException, IllegalAccessException {
         ResourceBundle.clearCache(urlClassLoader);
-        if (instantiator != null) {
-            Field field = DefaultInstantiator.class
-                    .getDeclaredField("i18nProvider");
-            field.setAccessible(true);
-            ((AtomicReference<I18NProvider>) field.get(instantiator)).set(null);
-            field.setAccessible(false);
-        }
+        I18NProviderTest.clearI18NProviderField();
     }
 
     @Test
@@ -90,9 +82,8 @@ public class DefaultInstantiatorI18NTest {
             VaadinService service = Mockito.mock(VaadinService.class);
             mockLookup(service);
 
-            instantiator = new DefaultInstantiator(service);
-
-            I18NProvider i18NProvider = instantiator.getI18NProvider();
+            I18NProvider i18NProvider = new DefaultInstantiator(service)
+                    .getI18NProvider();
             Assert.assertNotNull(i18NProvider);
             Assert.assertTrue(i18NProvider instanceof DefaultI18NProvider);
 
@@ -134,9 +125,8 @@ public class DefaultInstantiatorI18NTest {
             VaadinService service = Mockito.mock(VaadinService.class);
             mockLookup(service);
 
-            instantiator = new DefaultInstantiator(service);
-
-            I18NProvider i18NProvider = instantiator.getI18NProvider();
+            I18NProvider i18NProvider = new DefaultInstantiator(service)
+                    .getI18NProvider();
             Assert.assertNotNull(i18NProvider);
             Assert.assertTrue(i18NProvider instanceof DefaultI18NProvider);
 
@@ -173,9 +163,8 @@ public class DefaultInstantiatorI18NTest {
             VaadinService service = Mockito.mock(VaadinService.class);
             mockLookup(service);
 
-            instantiator = new DefaultInstantiator(service);
-
-            I18NProvider i18NProvider = instantiator.getI18NProvider();
+            I18NProvider i18NProvider = new DefaultInstantiator(service)
+                    .getI18NProvider();
             Assert.assertNotNull(i18NProvider);
             Assert.assertTrue(i18NProvider instanceof DefaultI18NProvider);
 

--- a/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/i18n/DefaultInstantiatorI18NTest.java
@@ -52,7 +52,8 @@ public class DefaultInstantiatorI18NTest {
     private ClassLoader urlClassLoader;
 
     @Before
-    public void init() throws IOException {
+    public void init()
+            throws IOException, NoSuchFieldException, IllegalAccessException {
         File resources = temporaryFolder.newFolder();
 
         translations = new File(resources, DefaultI18NProvider.BUNDLE_FOLDER);
@@ -60,6 +61,7 @@ public class DefaultInstantiatorI18NTest {
 
         urlClassLoader = new URLClassLoader(
                 new URL[] { resources.toURI().toURL() });
+        I18NProviderTest.clearI18NProviderField();
     }
 
     @After

--- a/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
@@ -54,8 +54,7 @@ public class I18NProviderTest {
 
     @Test
     public void property_defined_should_init_registy_with_provider()
-            throws ServletException, ServiceException, NoSuchFieldException,
-            IllegalAccessException {
+            throws ServletException, ServiceException {
         config.setApplicationOrSystemProperty(InitParameters.I18N_PROVIDER,
                 TestProvider.class.getName());
 
@@ -65,18 +64,11 @@ public class I18NProviderTest {
                 .getInstantiator();
         Assert.assertEquals("Found wrong registry", TestProvider.class,
                 instantiator.getI18NProvider().getClass());
-
-        Field field = DefaultInstantiator.class
-                .getDeclaredField("i18nProvider");
-        field.setAccessible(true);
-        ((AtomicReference<I18NProvider>) field.get(instantiator)).set(null);
-        field.setAccessible(false);
     }
 
     @Test
     public void with_defined_provider_locale_should_be_the_available_one()
-            throws ServletException, ServiceException, NoSuchFieldException,
-            IllegalAccessException {
+            throws ServletException, ServiceException {
         config.setApplicationOrSystemProperty(InitParameters.I18N_PROVIDER,
                 TestProvider.class.getName());
 
@@ -90,17 +82,22 @@ public class I18NProviderTest {
         Assert.assertEquals("Locale was not the defined locale",
                 i18NProvider.getProvidedLocales().get(0),
                 VaadinSession.getCurrent().getLocale());
-
-        Field field = DefaultInstantiator.class
-                .getDeclaredField("i18nProvider");
-        field.setAccessible(true);
-        ((AtomicReference<I18NProvider>) field.get(instantiator)).set(null);
-        field.setAccessible(false);
     }
 
     @After
-    public void clearCurrentInstances() {
+    public void clearCurrentInstances()
+            throws NoSuchFieldException, IllegalAccessException {
         CurrentInstance.clearAll();
+        clearI18NProviderField();
+    }
+
+    public static void clearI18NProviderField()
+            throws NoSuchFieldException, IllegalAccessException {
+        Field field = DefaultInstantiator.class
+                .getDeclaredField("i18nProvider");
+        field.setAccessible(true);
+        ((AtomicReference<I18NProvider>) field.get(null)).set(null);
+        field.setAccessible(false);
     }
 
     private void initServletAndService(DeploymentConfiguration config)

--- a/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
@@ -24,8 +24,10 @@ import java.util.Locale;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.ReentrantLock;
 
+import net.jcip.annotations.NotThreadSafe;
 import org.junit.After;
 import org.junit.Assert;
+import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mockito;
 
@@ -37,6 +39,7 @@ import com.vaadin.flow.internal.CurrentInstance;
 import com.vaadin.flow.shared.ApplicationConstants;
 import com.vaadin.tests.util.MockDeploymentConfiguration;
 
+@NotThreadSafe
 public class I18NProviderTest {
 
     private VaadinServletService service;
@@ -53,7 +56,7 @@ public class I18NProviderTest {
     }
 
     @Test
-    public void property_defined_should_init_registy_with_provider()
+    public void property_defined_should_init_registry_with_provider()
             throws ServletException, ServiceException {
         config.setApplicationOrSystemProperty(InitParameters.I18N_PROVIDER,
                 TestProvider.class.getName());
@@ -82,6 +85,12 @@ public class I18NProviderTest {
         Assert.assertEquals("Locale was not the defined locale",
                 i18NProvider.getProvidedLocales().get(0),
                 VaadinSession.getCurrent().getLocale());
+    }
+
+    @Before
+    public void initState()
+            throws NoSuchFieldException, IllegalAccessException {
+        clearI18NProviderField();
     }
 
     @After

--- a/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/I18NProviderTest.java
@@ -19,7 +19,9 @@ import jakarta.servlet.ServletException;
 import jakarta.servlet.http.HttpServletRequest;
 import jakarta.servlet.http.HttpSession;
 
+import java.lang.reflect.Field;
 import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.ReentrantLock;
 
 import org.junit.After;
@@ -52,33 +54,48 @@ public class I18NProviderTest {
 
     @Test
     public void property_defined_should_init_registy_with_provider()
-            throws ServletException, ServiceException {
+            throws ServletException, ServiceException, NoSuchFieldException,
+            IllegalAccessException {
         config.setApplicationOrSystemProperty(InitParameters.I18N_PROVIDER,
                 TestProvider.class.getName());
 
         initServletAndService(config);
 
+        Instantiator instantiator = VaadinService.getCurrent()
+                .getInstantiator();
         Assert.assertEquals("Found wrong registry", TestProvider.class,
-                VaadinService.getCurrent().getInstantiator().getI18NProvider()
-                        .getClass());
+                instantiator.getI18NProvider().getClass());
+
+        Field field = DefaultInstantiator.class
+                .getDeclaredField("i18nProvider");
+        field.setAccessible(true);
+        ((AtomicReference<I18NProvider>) field.get(instantiator)).set(null);
+        field.setAccessible(false);
     }
 
     @Test
     public void with_defined_provider_locale_should_be_the_available_one()
-            throws ServletException, ServiceException {
+            throws ServletException, ServiceException, NoSuchFieldException,
+            IllegalAccessException {
         config.setApplicationOrSystemProperty(InitParameters.I18N_PROVIDER,
                 TestProvider.class.getName());
 
         initServletAndService(config);
 
-        I18NProvider i18NProvider = VaadinService.getCurrent().getInstantiator()
-                .getI18NProvider();
+        Instantiator instantiator = VaadinService.getCurrent()
+                .getInstantiator();
+        I18NProvider i18NProvider = instantiator.getI18NProvider();
         Assert.assertNotNull("No provider for ", i18NProvider);
 
         Assert.assertEquals("Locale was not the defined locale",
                 i18NProvider.getProvidedLocales().get(0),
                 VaadinSession.getCurrent().getLocale());
 
+        Field field = DefaultInstantiator.class
+                .getDeclaredField("i18nProvider");
+        field.setAccessible(true);
+        ((AtomicReference<I18NProvider>) field.get(instantiator)).set(null);
+        field.setAccessible(false);
     }
 
     @After

